### ISSUE TITLE
Deadlock in NSRE code

### ISSUE THREAD

**Issue by jake-maloney** [OpenTSDB/asynchbase#163]
We are currently seeing the following deadlock in 1.7.2.

The "extra" lock seems to have been added with this fix https://github.com/OpenTSDB/asynchbase/commit/5653d520130035744a3f43434d66ebe2bc8ce197

```
Found one Java-level deadlock:
=============================
"AsyncHBase Timer RPC Timeout Timer #3":
  waiting to lock monitor 0x00007f9ec4014e78 (object 0x00007fa750eb6448, a org.hbase.async.GetRequest),
  which is held by "AsyncHBase I/O Worker #2"
"AsyncHBase I/O Worker #2":
  waiting to lock monitor 0x00007fa184012848 (object 0x00007faa04fe1568, a java.util.ArrayList),
  which is held by "AsyncHBase Timer RPC Timeout Timer #3"

Java stack information for the threads listed above:
===================================================
"AsyncHBase Timer RPC Timeout Timer #3":
	at org.hbase.async.HBaseClient.handleNSRE(HBaseClient.java:2804)
	- waiting to lock <0x00007fa750eb6448> (a org.hbase.async.GetRequest)
	at org.hbase.async.HBaseClient.sendRpcToRegion(HBaseClient.java:1957)
	at org.hbase.async.HBaseClient$1RetryNSREd.call(HBaseClient.java:2907)
	- locked <0x00007faa04fe1568> (a java.util.ArrayList)
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1261)
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1240)
	at com.stumbleupon.async.Deferred.callback(Deferred.java:988)
	at org.hbase.async.HBaseRpc.callback(HBaseRpc.java:712)
	at org.hbase.async.HBaseClient.handleNSRE(HBaseClient.java:2816)
	- locked <0x00007fa750eb64d8> (a org.hbase.async.GetRequest)
	at org.hbase.async.HBaseClient.sendRpcToRegion(HBaseClient.java:1957)
	at org.hbase.async.HBaseClient$1RetryNSREd.call(HBaseClient.java:2876)
	- locked <0x00007faa04baf9f0> (a java.util.ArrayList)
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1261)
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1240)
	at com.stumbleupon.async.Deferred.callback(Deferred.java:988)
	at org.hbase.async.HBaseRpc.callback(HBaseRpc.java:712)
	at org.hbase.async.HBaseRpc$TimeoutTask.run(HBaseRpc.java:632)
	at org.jboss.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:556)
	at org.jboss.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:632)
	at org.jboss.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:369)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at java.lang.Thread.run(Thread.java:745)
"AsyncHBase I/O Worker #2":
	at org.hbase.async.HBaseClient$1RetryNSREd.call(HBaseClient.java:2867)
	- waiting to lock <0x00007faa04fe1568> (a java.util.ArrayList)
	- locked <0x00007faa04baf928> (a java.util.ArrayList)
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1261)
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1240)
	at com.stumbleupon.async.Deferred.callback(Deferred.java:988)
	at org.hbase.async.HBaseRpc.callback(HBaseRpc.java:712)
	at org.hbase.async.HBaseClient.handleNSRE(HBaseClient.java:2816)
	- locked <0x00007fa750eb6448> (a org.hbase.async.GetRequest)
	at org.hbase.async.RegionClient.decode(RegionClient.java:1527)
	at org.hbase.async.RegionClient.decode(RegionClient.java:88)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:500)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.hbase.async.RegionClient.handleUpstream(RegionClient.java:1223)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.SimpleChannelHandler.messageReceived(SimpleChannelHandler.java:142)
	at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88)
	at org.jboss.netty.handler.timeout.IdleStateAwareChannelHandler.handleUpstream(IdleStateAwareChannelHandler.java:36)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.handler.timeout.IdleStateHandler.messageReceived(IdleStateHandler.java:294)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.hbase.async.HBaseClient$RegionClientPipeline.sendUpstream(HBaseClient.java:3121)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:318)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

Found 1 deadlock.```

**Comment by jasonculverhouse** [comment]
FYI this fixes the issue https://github.com/OpenTSDB/asynchbase/pull/159/files

**Comment by jake-maloney** [comment]
So there is a "trivial" fix to this problem and thats to remove the added synchronization

```
synchronized (exists_rpc) {
...
}
```

However, this isn't technically thread safe and I would say to make it thread-safe without a synchronization would require some changes to the model. For example, if we were to make the fields in question volatile or add some sort of lock to HBaseRpc we are introducing synchronization into this class, whose java doc specifically says `Implementations of this class are <b>not</b> expected to be synchronized.`. I also think fundamentally part of the problem is the nested `synchronized` blocks all around the client which make it very prone to this type of issue especially when you have io threads and the timer thread both handling the NSRE code path.

So there are a few options:
1) I can submit the patch without the thread-safety. 
2) I can change the HBaseRpc and break that contract. 
3) I can rework the model/client entirely.
4) Someone else can figure out the nested locking in a way that works.

What do the maintainers want? How can I assist in fixing this (sooner rather than later)?

**Comment by manolama** [comment]
Hmm, I think #1 is ok here for now. Maybe sync only for the call to ``exists_rpc.setSuspendedProbe(false);``. Thanks for finding and fixing!

**Comment by manolama** [comment]
Hmm, that'd be better, or we can at least sync each RPC on something that doesn't change, e.g. the ``key``.

**Comment by manolama** [comment]
Merged in 5e6861d3ca018254070a1c91f654c4a021949d43, lets re-open if this continues please. Thanks!
